### PR TITLE
Use Google Oauth2 library

### PR DESCRIPTION
**Pull request by kostjas**

### CODE REVIEW

File: project/build.properties [hunk 1]
```diff
@@ -14,4 +14,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
-sbt.version=1.4.7
+sbt.version=1.5.6
```
Comment (SuperIzya) [inline project/build.properties:17]: FYI: The latest version is `1.6.1` now

File: src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala [hunk 2]
```diff
@@ -13,21 +13,19 @@
...
+  override def createRelation(sqlContext: SQLContext, parameters: Map[String, String]): BaseRelation = {
     createRelation(sqlContext, parameters, null)
   }
 
-  private[spreadsheets] def pathToSheetNames(parameters: Map[String, String]): (String, String) = {
-    val path = parameters.getOrElse("path", sys.error("'path' must be specified for spreadsheets."))
+  private[spreadsheets] def pathToSheetNames(path: String): (String, String) = {
     val elems = path.split('/')
     if (elems.length < 2)
       throw new Exception("'path' must be formed like '<spreadsheet>/<worksheet>'")
```
Comment (SuperIzya) [inline src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala:31]: Yuk! Throwing exceptions!!!

File: src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala [hunk 3]
```diff
@@ -36,27 +34,31 @@ class DefaultSource extends RelationProvider with SchemaRelationProvider with Cr
...
   override def createRelation(sqlContext: SQLContext, mode: SaveMode, parameters: Map[String, String], data: DataFrame): BaseRelation = {
-    val (spreadsheetName, worksheetName) = pathToSheetNames(parameters)
-    implicit val context = createSpreadsheetContext(parameters)
+    val path = parameters.getOrElse("path", sys.error("'path' must be specified for spreadsheets."))
+    val (spreadsheetName, worksheetName) = pathToSheetNames(path)
+    implicit val context: SparkSpreadsheetService.SparkSpreadsheetContext = createSpreadsheetContext(parameters)
     val spreadsheet = SparkSpreadsheetService.findSpreadsheet(spreadsheetName)
-    if(!spreadsheet.isDefined)
-      throw new RuntimeException(s"no such a spreadsheet: $spreadsheetName")
+      .fold(throw new RuntimeException(s"no such a spreadsheet: $spreadsheetName"))(identity)
```
Comment (SuperIzya) [inline src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala:49]: Yuk!

File: src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala [hunk 4]
```diff
@@ -36,27 +34,31 @@ class DefaultSource extends RelationProvider with SchemaRelationProvider with Cr
...
 
-  private[spreadsheets] def createSpreadsheetContext(parameters: Map[String, String]) = {
-    val serviceAccountIdOption = parameters.get("serviceAccountId")
-    val credentialPath = parameters.getOrElse("credentialPath", DEFAULT_CREDENTIAL_PATH)
-    SparkSpreadsheetService(serviceAccountIdOption, new File(credentialPath))
+  private[spreadsheets] def createSpreadsheetContext(parameters: Map[String, String]): SparkSpreadsheetService.SparkSpreadsheetContext = {
+    val credentials = (parameters.get("credentialsPath"), parameters.get("credentialsJson")) match {
+      case (Some(credentialsPath), None) => Credentials.credentialsFromFile(credentialsPath)
+      case (None, Some(credentialsJson)) => Credentials.credentialsFromJsonString(credentialsJson)
+      case _ => throw new IllegalStateException("Either 'credentialsPath' or 'credentialsJson' parameter should be defined.")
```
Comment (SuperIzya) [inline src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala:59]: Yuk!

File: src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala [hunk 5]
```diff
@@ -36,27 +34,31 @@ class DefaultSource extends RelationProvider with SchemaRelationProvider with Cr
...
+    }
+    SparkSpreadsheetService(credentials)
   }
 
   private[spreadsheets] def createRelation(sqlContext: SQLContext,
```
Comment (SuperIzya) [inline src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala:64]: Can't comment in proper place (don't know why). But:
```
if (schema == null) {
      createRelation(sqlContext, context, spreadsheetName, worksheetName, None)
}
else {
      createRelation(sqlContext, context, spreadsheetName, worksheetName, Some(schema))
}
```
can be rewritten as 
```
createRelation(sqlContext, context, spreadsheetName, worksheetName, Option(schema))
```

File: src/main/scala/com/github/potix2/spark/google/spreadsheets/SparkSpreadsheetService.scala [hunk 6]
```diff
@@ -169,14 +151,15 @@ object SparkSpreadsheetService {
     def name: String = sheet.getProperties.getTitle
     lazy val values = {
       val valueRange = context.query(spreadsheet.getSpreadsheetId, name)
-      if ( valueRange.getValues != null )
+      if ( valueRange.getValues != null ) {
```
Comment (SuperIzya) [inline src/main/scala/com/github/potix2/spark/google/spreadsheets/SparkSpreadsheetService.scala:154]: IMHO `Option(valueRange.getValues).fold(...)` will be nicer.

File: src/main/scala/com/github/potix2/spark/google/spreadsheets/SparkSpreadsheetService.scala [hunk 7]
```diff
@@ -251,6 +234,7 @@ object SparkSpreadsheetService {
    * @param context
    * @return
    */
-  def findSpreadsheet(spreadsheetName: String)(implicit context: SparkSpreadsheetContext): Option[SparkSpreadsheet] =
-    Some(context.findSpreadsheet(spreadsheetName))
+  def findSpreadsheet(spreadsheetName: String)
+                     (implicit context: SparkSpreadsheetContext): Option[SparkSpreadsheet] =
```
Comment (SuperIzya) [inline src/main/scala/com/github/potix2/spark/google/spreadsheets/SparkSpreadsheetService.scala:238]: Does it have to be implicit?

**Review by SuperIzya — COMMENTED** [review]
Do you really need all this functionality to manipulate the spreadsheets?

File: src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala (hunk 2 above)
Comment (kostjas) [reply src/main/scala/com/github/potix2/spark/google/spreadsheets/DefaultSource.scala:30]: That was present in the original repository. This should be fixed in a separate PR.

File: src/main/scala/com/github/potix2/spark/google/spreadsheets/SparkSpreadsheetService.scala (hunk 6 above)
Comment (kostjas) [reply src/main/scala/com/github/potix2/spark/google/spreadsheets/SparkSpreadsheetService.scala]: You are right! Original types didn't reflect real data. This code was refactored

File: build.sbt [hunk 8]
```diff
@@ -12,19 +12,17 @@ description := "Google Spreadsheets datasource for SparkSQL and DataFrames."
...
   "org.scalatest" %% "scalatest" % "3.2.9" % "test",
-  ("com.google.api-client" % "google-api-client" % "1.31.5").
-    exclude("com.google.guava", "guava-jdk5")
-    .exclude("com.google.guava", "guava:30.1.1-android"),
-  "com.google.apis" % "google-api-services-sheets" % "v4-rev20210322-1.31.0"
+  "com.google.apis" % "google-api-services-sheets" % "v4-rev20210629-1.32.1",
+  "com.google.auth" % "google-auth-library-oauth2-http" % "1.4.0"
 )
 
 libraryDependencies ++= Seq(
```
[the remark this answers is not in the document]
Comment (kostjas) [reply build.sbt]: It was the config from the original repository.  I didn't touch them.
The config was refactored.

**Pull request merged**